### PR TITLE
Improvements: More general to and from methods

### PR DESCRIPTION
**Pull request by andrew-causalens**
## Description and Motivation

Updated to/from methods to ensure the type matches the calling class.  There is also a small bug fix where the node variable type was not passed to the skeleton node.

## Additional Context

Part of a wider effort to improve the generalisation of the base `CausalGraph` class.

## How Has This Been Tested?

Extended unit tests. Also added some tests for JSON serialisability of `Skeleton`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [ ] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [x] I have implemented all requirements? (see JIRA, project documentation)
- [x] I am affecting someone else's work? If so: I have added those people as reviewers?
- [x] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [x] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [x] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [x] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: cai_causal_graph/causal_graph.py [hunk 1]
```diff
@@ -104,7 +104,9 @@ def __iter__(self) -> Iterator:
     def nodes(self) -> List[Node]:
         """Return a list of all nodes."""
         # Instantiate new nodes to remove any edge information.
-        return [Node(n.identifier, meta=n.meta) for n in self._graph.nodes]
+        return [
+            self._graph._NodeCls(n.identifier, meta=n.meta, variable_type=n.variable_type) for n in self._graph.nodes
```
Comment (andrew-causalens) [inline cai_causal_graph/causal_graph.py:108]: This is a bug fix as before `variable_type` was not being passed in so skeleton would not know it was a binary node for example. But I also changed to use the graph node class so the skeleton will have a time series node, as opposed to standard node, if it is a time series graph.

File: cai_causal_graph/causal_graph.py [hunk 2]
```diff
@@ -207,13 +209,32 @@ def adjacency_matrix(self) -> numpy.ndarray:
...
-        """Instantiate a `cai_causal_graph.causal_graph.Skeleton` object from an adjacency matrix."""
-        graph: CausalGraph = CausalGraph.from_adjacency_matrix(adjacency=adjacency, node_names=node_names)
-        return Skeleton(graph=graph)
+        """
+        Instantiate a `cai_causal_graph.causal_graph.Skeleton` object from an adjacency matrix.
+
+        :param adjacency: The adjacency matrix.
+        :param node_names: A list of the node names. These should be in the order of the rows and columns of the
+            adjacency matrix.
+        :param graph_class: The appropriate subclass of `cai_causal_graph.causal_graph.CausalGraph` to use. If `None`,
```
Comment (andrew-causalens) [inline cai_causal_graph/causal_graph.py:225]: @maxelliott-causalens this is the only bit where I am not sure if it is a good solution. Only affects the `from` methods for `Skeleton`.

File: cai_causal_graph/graph_components.py [hunk 3]
```diff
@@ -48,7 +48,7 @@ def __init__(
         self._identifier = identifier
 
         self.variable_type: NodeVariableType = variable_type
-        self.meta = dict() if meta is None else meta
+        self.meta = dict() if meta is None else deepcopy(meta)
```
Comment (ilya-causalens) [inline cai_causal_graph/graph_components.py:51]: Why do we deepcopy here?

**Review by maxelliott-causalens — CHANGES_REQUESTED** [review]
Thanks @andrew-causalens for sorting this! I've requested changes as quite a bit of code can (and should) now be removed from TimeSeriesCausalGraph. Removing them also means we'll get some testing for the new `classmethods` for free as there are checks for TSCG checking that the methods will return the correct type.

File: cai_causal_graph/graph_components.py (hunk 3 above)
Comment (andrew-causalens) [reply cai_causal_graph/graph_components.py:51]: Added to align with `TimeSeriesNode` but will remove the copy of meta in both places.

File: cai_causal_graph/causal_graph.py (hunk 2 above)
Comment (andrew-causalens) [reply cai_causal_graph/causal_graph.py:225]: I don't think we can as we would want to have `TimeSeriesNode` as the node class within the `Skeleton` of a `TimeSeriesCausalGraph`.

File: cai_causal_graph/time_series_causal_graph.py [hunk 4]
```diff
@@ -118,6 +119,15 @@ def __init__(
...
         self._minimal_graph: Optional[TimeSeriesCausalGraph] = None
 
+    # Overwrite type annotations for linting and code completion.
+    # from_adjacency_matrix: Callable[
+    #     [Arg(numpy.ndarray, 'adjacency'), DefaultArg(Optional[List[Union[NodeLike, int]]], 'node_names')],
+    #     TimeSeriesCausalGraph,
+    # ]
+    # from_skeleton: Callable[[Arg(Skeleton, 'skeleton')], TimeSeriesCausalGraph]
+    # from_networkx: Callable[[Arg(networkx.Graph, 'g')], TimeSeriesCausalGraph]
+    # from_gml_string: Callable[[Arg(str, 'gml')], TimeSeriesCausalGraph]
```
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:129]: @maxelliott-causalens I can't actually get this to work per the spec. Linting passes without it.

**Review by maxelliott-causalens — APPROVED** [review]
Will approve to unblock for when you fix the issues!

**Pull request merged**